Quote pattern matching runtime spec

nicolasstucki opened the pull request:
A first draft of the evaluation semantics of quote pattern matching.

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+
+Bellow we provide an informal description of the pattern evailatuion.
+We use `'{..}` for expression, `'[..]` for types and `{{..}}` for patterns nested in expressions.
+The semantics are defined as a list of reduction rules that are tried one by one until one matches.
```
It would be nice to first define the syntax of the pattern language.

nicolasstucki replied:
That will be the next step

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+* `withEnv(x1 -> y1, ..., xn -> yn)(matching)` evaluates mathing recording that `xi` is equivalent to `yi`.
+* `matched` denotes that the the match succedded and `matched('{e})` denotes that a matech succeded and extracts `'{e}`
+* `&&&` matches if both sides match. Concatenates the extracted expressions of both sides.
+
```
Typesetting: it would be nicer to typeset the semantics as inference rules.

Note that the semantics should not only tell whether match or not, but also a map of bindings.

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
````diff
@@ -0,0 +1,64 @@
+* `&&&` matches if both sides match. Concatenates the extracted expressions of both sides.
+
+```scala
+'{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>   matched('{e})
````
This looks like to be a wildcard pattern, the type checking looks redundant.

The condition `isColosedUnder()('{e}) ` is a little controversial. I see the motivation is for safety. The question is:

- whether it is indeed safe
- potentially high usability penalty here, which may out-weigh the concern of safety

Blaisorblade replied:
`matched('{e})` probably also means "match `x` to `e`", right?

nicolasstucki replied:
Actually `$` should not be part of this part of the specification, here we deal with `hole[T]` which is the thing that is generated after typing the quote pattern.

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
````diff
@@ -0,0 +1,64 @@
+
+```scala
+'{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>   matched('{e})
+'{ e } =?= '{ $y(x1, ..., xn) }  &&  isColosedUnder(x1, ... xn)('{e})  ===>   matched(lift(x1, ..., xn)('{e}))
````
This quoted pattern is a little counter-intuitive, both syntactically and semantically. I'm not sure how useful and usable it is in practice.

LPTK replied:
It's a higher-order pattern variable (see https://dl.acm.org/doi/10.1145/3136040.3136043, section 5.7). I think it's quite intuitive in actual examples. But it's a pretty limited pattern, and as was discussed before in the group meeting, it could be better to go with a scope-tracking type system instead.

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ e } =?= '{ $y(x1, ..., xn) }  &&  isColosedUnder(x1, ... xn)('{e})  ===>   matched(lift(x1, ..., xn)('{e}))
+
+'{ lit } =?= '{ lit }   ===>   matched
+'{ e: T } =?= '{ p: P }   ===>   '{e} =?= '{p} &&& '[T] =?= '[P]
```
This no longer agrees with #8684

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+
+'{ lit } =?= '{ lit }   ===>   matched
+'{ e: T } =?= '{ p: P }   ===>   '{e} =?= '{p} &&& '[T] =?= '[P]
+'{ e } =?= '{ p: P }   ===>   '{e} =?= '{p}
```
Is it not obvious that ignoring the type test `P` is safe?

Blaisorblade replied:
Agreed; I'd have expected something like "take the type `T` of `e` and check `T <:< P`, unifying type variables in `P` as needed".

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ e.x } =?= '{ p.x }   ===>   '{e} =?= '{p}
+'{ x } =?= '{ x }   ===>   matched
+'{e0.f(e1, ..., en)} =?= '{p0.f(p1, ..., p2)}   ===>   '{e0} =?= '{p0} &&& '{e1} =?= '{p1} &&& ... %% '{en} =?= '{pn}
+'{e.f[T1, ..., Tn]} =?= '{p.f[P1, ..., Pn]}   ===>   '{e} =?= '{p} &&& '[T1] =?= '{P1} &&& ... %% '[Tn] =?= '[Pn]
```
It seems we miss a case for `e.f[T1, ..., Tn](e1, ..., en)`, which subsumes the two cases above.

Blaisorblade replied:
Multiple choices are possible, but it seems these two cases match respectively `Apply` and `TypeApply` — and since those are distinct constructors in Tasty, they should be distinct cases in the semantics for consistency, so that each clause matches exactly one term constructor.

LPTK replied:
The problem is that the `e0.f(e1, ..., en)` rule should really be `e0(e1, ..., en)` or else it disallows expressions of the form `x.f[...](e1, ..., en)`.

nicolasstucki replied:
Indeed, that is what the implementation does

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ e0 = e1 } =?= '{ p0 = p1 } && '{e0} =?= '{p0}   ===>   '{e1} =?= '{p1}
+'{ new T } =?= '{ new T }   ===>   matched
+'{ this } =?= '{ this }   ===>   matched
+'{ e.super[T] } =?= '{ p.super[T] }   ===>   '{e} =?= '{p}
```
`e.super[T](e1, ..., en)`

Blaisorblade replied:
This is a `Super` node, so same reasoning.

nicolasstucki replied:
Should have been `'{ e.super }` and the application rules take care of the arguments

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ new T } =?= '{ new T }   ===>   matched
+'{ this } =?= '{ this }   ===>   matched
+'{ e.super[T] } =?= '{ p.super[T] }   ===>   '{e} =?= '{p}
+'{ Seq(e0, ..., en): _* } =?= '{ Seq(p0, ..., pn): _* }   ===>   '{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn}
```
Maybe change it to `'{ e: _* } ~ '{ p : _*} `

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'[T] =?= '[P] && T <:< P   ===>   matched
+'[ T0[T1, ..., Tn] ] =?= '[ P0[P1, ..., Pn] ]   ===>   '[T0] =?= '[P0] &&& ... &&& '[Tn] =?= '[Pn]
+'[T @annot] =?= '[P]   ===>   '[T] =?= '[P]
+'[T] =?= '[P @annot]   ===>   '[T] =?= '[P]
```
We need to define `P` first.

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ val x: T = e1; e2 } =?= '{ val y: P = p1; p2 }   ===>   withEnv(x -> y)('[T] =?= '[P] &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2})
+'{ def x0(x1: T1, ..., xn: Tn): T0 = e1; e2 } =?= '{ def y0(y1: P1, ..., yn: Pn): P0 = p1; p2 }   ===>   withEnv(x0 -> y0, ..., xn -> yn)('[T0] =?= '[P0] &&& ... &&& '[Tn] =?= '[Pn] &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2})
+
+'{ e1; e2 } =?= '{ p1; p2 }   ===>   '{e1} =?= '{p1} &&& '{e2} =?= '{p2}
```
Seems it can be removed, it's a special case of the block.

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'[T @annot] =?= '[P]   ===>   '[T] =?= '[P]
+'[T] =?= '[P @annot]   ===>   '[T] =?= '[P]
+
+'{ {e0; e1; ...; en}; em } =?= '{ {p0; p1; ...; pm}; em }   ===>   '{ e0; {e1; ...; en; em} } =?= '{ p0; {p1; ...; pm; em} }
```
```suggestion
'{ {e0; e1; ...; en}; em } =?= '{ {p0; p1; ...; pn}; pm }   ===>   '{ e0; {e1; ...; en; em} } =?= '{ p0; {p1; ...; pn; pm} }
```

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+
+'{ {e0; e1; ...; en}; em } =?= '{ {p0; p1; ...; pm}; em }   ===>   '{ e0; {e1; ...; en; em} } =?= '{ p0; {p1; ...; pm; em} }
+'{ val x: T = e1; e2 } =?= '{ val y: P = p1; p2 }   ===>   withEnv(x -> y)('[T] =?= '[P] &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2})
+'{ def x0(x1: T1, ..., xn: Tn): T0 = e1; e2 } =?= '{ def y0(y1: P1, ..., yn: Pn): P0 = p1; p2 }   ===>   withEnv(x0 -> y0, ..., xn -> yn)('[T0] =?= '[P0] &&& ... &&& '[Tn] =?= '[Pn] &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2})
```
The semantics for local definitions is surprising. I see the intention is to avoid scope extrusion, but it seems to be a big usability penalty here.

LPTK replied:
@liufengyun What alternative semantics are you thinking about?

liufengyun replied:
The original comment is actually about the following rule:

> '{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>  
 matched('{e})

From my limited meta-programming experience, I like quoted pattern matching on terms and types a lot, it's a big productivity booster. In particular, it saves a lot of boilerplate when pattern matching on known methods and types. This makes developing macro-based DSLs a joyful experience. Previously, they have to be performed via low-level reflect API in a tedious way or doing it in an unsafe way via string comparison.  @nicolasstucki has done a great job here for Scala meta-programmers👍 

However, I'm not sure that making it _safer_ and feature-complete will help much in practice, in particular when it comes at the price of usability. By usability, I mean simple API, natural and precise semantics, stability, debuggability, a clear-cut boundary of what is supported (accountable & trustable) and what is not.

For example, maybe it is safe drop the support for quoted code patterns of pattern matches and definitions, and leave them to reflect API?

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ e1; e2 } =?= '{ p1; p2 }   ===>   '{e1} =?= '{p1} &&& '{e2} =?= '{p2}
+
+'{ e0 match { case u1 => e1; ...; case un => en } } =?= '{ p0 match { case q1 => p1; ...; case qn => pn } }   ===>
+'{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn} &&& '{{u1}} =?= '{{q1}} &&& ... &&& '{{un}} =?= '{{qn}}
```
I'm not sure this is useful in practice. Maybe disallow pattern matching in quoted pattern, much fewer bugs/crashes.

nicolasstucki replied:
It is a good idea to disallow these for now

liufengyun commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ e0 match { case u1 => e1; ...; case un => en } } =?= '{ p0 match { case q1 => p1; ...; case qn => pn } }   ===>
+'{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn} &&& '{{u1}} =?= '{{q1}} &&& ... &&& '{{un}} =?= '{{qn}}
+
+'{ try e0 catch { case u1 => e1; ...; case un => en } finally ef } =?= '{ try p0 catch { case q1 => p1; ...; case qn => pn } finally pf }   ===>   '{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn} &&& '{{u1}} =?= '{{q1}} &&& ... &&& '{{un}} =?= '{{qn}} &&& '{ef} =?= '{pf}
```
Maybe leave the `catch` handler as an expression, and leave the pattern matching to Tasty Reflect.

Blaisorblade replied:
Is this about the spec or a feature request?
But yes, it'd certainly be nice to be able to match against unspecified `catch` clauses — looking at the definition of `Try`, it'd be nice to have list pattern matching (as an option) without resorting to Tasty Reflect.

Then you'd match against `{ try e0 catch c finally ef }` and have a separate grammar for `c` and corresponding patterns.

nicolasstucki replied:
Currently, the grammar for `c` uses the `'{{...}}` (instead of `'{...}`) to denote patterns.

Blaisorblade commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+
+## Evaluation sematics
+
+Bellow we provide an informal description of the pattern evailatuion.
```
Some short summary is probably a good idea, and apart from typos this seems good?

Blaisorblade commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+
+'{ try e0 catch { case u1 => e1; ...; case un => en } finally ef } =?= '{ try p0 catch { case q1 => p1; ...; case qn => pn } finally pf }   ===>   '{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn} &&& '{{u1}} =?= '{{q1}} &&& ... &&& '{{un}} =?= '{{qn}} &&& '{ef} =?= '{pf}
+
+'{{ _ }} =?= '{{ _ }}   ===>   matched
```
What's an `_` on the LHS? I guess you're allowing anything there, or is this matching an `Inferred` node?

Blaisorblade commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ e: T } =?= '{ p: P }   ===>   '{e} =?= '{p} &&& '[T] =?= '[P]
+'{ e } =?= '{ p: P }   ===>   '{e} =?= '{p}
+'{ e.x } =?= '{ p.x }   ===>   '{e} =?= '{p}
+'{ x } =?= '{ x }   ===>   matched
```
Should this be somewhere next with `'{{ `x` }} =?= '{{ `y` }}` , and maybe merged together?

Blaisorblade replied:
I now suspect I agree with this rule, but it took me a while to see why `'{ x } =?= '{ y }` is illegal (because alpha-equivalence is handled when checking local binders). Some explanation would help. (Who writes it is another question).

BTW, I'd hope `Select`s encoded as `Ident`s (per `desugarIdent`) show up as `Select` here. I see they do not in Tasty reflection... does at least -Ycheck ensure that they're represented consistently so it's well-defined what you have to write to match `Int` (as opposed to needing `Int`, `scala.Int` or `_root_.scala.Int` depending on, say, what the user wrote)?

Blaisorblade commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn} &&& '{{u1}} =?= '{{q1}} &&& ... &&& '{{un}} =?= '{{qn}}
+
+'{ try e0 catch { case u1 => e1; ...; case un => en } finally ef } =?= '{ try p0 catch { case q1 => p1; ...; case qn => pn } finally pf }   ===>   '{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn} &&& '{{u1}} =?= '{{q1}} &&& ... &&& '{{un}} =?= '{{qn}} &&& '{ef} =?= '{pf}
+
```
Please specify we're moving to the grammar of pattern quasiquotes...

Blaisorblade commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{{ e1 | ... | en }} =?= '{{ p1 | ... | pn }}   ===>   '{{e1}} =?= '{{p1}} &&& ... &&& '{{en}} =?= '{{pn}}
+'{{ e: T }} =?= '{{ p: U }}   ===>   '{{e}} =?= '{{p}} &&& '[T] =?= [U]
+'{{ e: T }} =?= '{{ p: U }}   ===>   '{{e}} =?= '{{p}} &&& '[T] =?= [U]
+'{{ `x` }} =?= '{{ `y` }}   ===>   matched
```
1) Why the nested backticks?
2) Why doesn't this bind `y` to `x`?

nicolasstucki replied:
Those mean that these are patterns.

I may have renamed them by mistake, I will double-check.

Blaisorblade commented:
Hi! I hope it's okay, I took the liberty of leaving a few review comments, also because this relates to #8600 (and I hope I can contribute a bit here).

smarter commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ if e0 then e1 else e2 } =?= '{ if p0 then p1 else p2 }   ===>  '{e0} =?= '{p0} &&& '{e1} =?= '{p1} &&& '{e2} =?= '{p2}
+'{ while e0 do e1 } =?= '{ while p0 do p1 }   ===>  '{e0} =?= '{p0} &&& '{e1} =?= '{p1}
+
+'{ e0 = e1 } =?= '{ p0 = p1 } && '{e0} =?= '{p0}   ===>   '{e1} =?= '{p1}
```
I find this a bit weird to read, if this is supposed to be a recursive definition for `=?=`, then I wouldn't expect multiple clause on the left of the arrow, instead I would expect:
```suggestion
'{ e0 = e1 } =?= '{ p0 = p1 } ===>   '{e0} =?= '{p0}  &&  '{e1} =?= '{p1}
```

nicolasstucki replied:
The current version reflects what the implementation does. I agree that we should have the second.

smarter commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+
+'[T] =?= '[P] && T <:< P   ===>   matched
+'[ T0[T1, ..., Tn] ] =?= '[ P0[P1, ..., Pn] ]   ===>   '[T0] =?= '[P0] &&& ... &&& '[Tn] =?= '[Pn]
+'[T @annot] =?= '[P]   ===>   '[T] =?= '[P]
```
Annotations should be handled by subtyping rules already (and some annotations can create proper subtype, see https://github.com/lampepfl/dotty/blob/4253708ee22265959b6768547169c330244bf9cb/compiler/src/dotty/tools/dotc/core/Types.scala#L4589)

smarter commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ e.super[T] } =?= '{ p.super[T] }   ===>   '{e} =?= '{p}
+'{ Seq(e0, ..., en): _* } =?= '{ Seq(p0, ..., pn): _* }   ===>   '{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn}
+
+'[T] =?= '[P] && T <:< P   ===>   matched
```
This seems to contradict the fact that scala.quoted.Type is invariant in its type parameter. I guess this is needed for https://github.com/lampepfl/dotty/pull/8695 to work, but as I noted there, I tihnk the proper thing to do is to create quoted type variable instead of doing type inference, picking "Any", and relying on this rule to get other types to match.

smarter commented on docs/docs/reference/metaprogramming/quote-matching.md:
```diff
@@ -0,0 +1,64 @@
+'{ Seq(e0, ..., en): _* } =?= '{ Seq(p0, ..., pn): _* }   ===>   '{e0} =?= '{p0} &&& ... &&& '{en} =?= '{pn}
+
+'[T] =?= '[P] && T <:< P   ===>   matched
+'[ T0[T1, ..., Tn] ] =?= '[ P0[P1, ..., Pn] ]   ===>   '[T0] =?= '[P0] &&& ... &&& '[Tn] =?= '[Pn]
```
> As in, will it successfully check List <: Seq

Yes.

> What happens if the variances differ?

We take variance into account: https://github.com/lampepfl/dotty/blob/63ee630044b5555df691de02e59fb6d43d121a6d/compiler/src/dotty/tools/dotc/core/TypeComparer.scala#L594-L595 (but note that type lambdas themselves have inferred variance now: https://github.com/lampepfl/dotty/pull/8082)

LPTK commented:
On the same topic, since it's related: I stumbled upon the `flops-rewrite` tests in `tests/run-macros/` and noticed that their encoding of safe term transformations, based on quote-matching, is unsound.

For instance, in [`tests/run-macros/flops-rewrite-3/Macro_1.scala`](https://github.com/lampepfl/dotty/blob/master/tests/run-macros/flops-rewrite-3/Macro_1.scala), if we try:

```scala
  val rewriter = Rewriter().withFixPoint.withPost(
    Transformation.safe[Any] {
      case _ => '{"oops"}
    }
```

We get a match error in the inner match of:

```scala
  def apply[T: Type](e: Expr[T])(using QuoteContext): Expr[T] = {
    e match {
      case '{ $e: U } => transform.applyOrElse(e, identity) match { case '{ $e2: T } => e2 }
      case e => e
    }
```

Indeed, the transformation may rewrite to something that's larger than the expected type, in which case it should not be deemed applicable. There is a `T <: U` check missing.

This is an important gotcha to watch out for, if a `rewrite` primitive is ever added to the reflection API.

nicolasstucki commented:
I've moved the spec to the `Matcher` class to make it easier to maintain in sync with the implementation. I also named the rules and added references to the rules in the implementation.

I improved the readability of the rules based on the feedback (still to be improved) as well as fixing some rules that were not in line with the implementation. The current state represented the current implementation.

For actual semantic issues, I did not fix the rules but added some TODOs in the code. Once we have this spec in we can fix those while updating the spec at the same time.

I suggest merging this and have any discussion on semantics on new issues that reference the spec.

Thank you for all the feedback.

nicolasstucki commented:
@liufengyun, could you re-review this. We should just make sure that the current spec is inline with the current implementation for this PR.

liufengyun commented on library/src/scala/internal/quoted/Matcher.scala:
```diff
@@ -4,6 +4,135 @@ import scala.annotation.internal.sharable
+ *   '{ new T } =?= '{ new T }   ===>   matched
+ *
+ *   /* Match this */
+ *   '{ this } =?= '{ this }   ===>   matched
```
```suggestion
 *   '{ C.this } =?= '{ C.this }   ===>   matched
```
